release/nightly_tests/dataset/multi_node_train_benchmark.py

```diff
@@ -205,32 +238,106 @@ def train_loop_per_worker():
 
     world_size = ray.train.get_context().get_world_size()
     all_workers_time_list_across_epochs = []
-    for i in range(args.num_epochs):
-        print(f"Epoch {i+1} of {args.num_epochs}")
-        num_rows = 0
-        start_t = time.time()
-
+    validation_accuracy_per_epoch = []
+    # Validation loop with non-random cropped dataset
+    # is only supported for image dataset.
+    run_validation_set = not args.synthetic_data and args.file_type == "image"
+    for epoch in range(args.num_epochs):
+        if args.synthetic_data:
+            # Generate a random batch, and continuously yield the same batch 1000 times.
+            sample_batch = {
+                "image": torch.rand(
+                    (args.batch_size, 3, DEFAULT_IMAGE_SIZE, DEFAULT_IMAGE_SIZE),
+                    device=device,
+                ),
+                "label": torch.randint(0, 1000, (args.batch_size,), device=device),
+            }
+            it = itertools.repeat(sample_batch, 1000)
+            batch_iter = it
         # Ray Data needs to call iter_torch_batches on each epoch.
-        if isinstance(it, ray.data.iterator.DataIterator):
+        elif isinstance(it, ray.data.iterator.DataIterator):
             batch_iter = it.iter_torch_batches(
                 batch_size=args.batch_size,
+                local_shuffle_buffer_size=args.batch_size / 2,
             )
 
-        print_at_interval = 1000
-        print_at = print_at_interval
-        for batch in batch_iter:
+            if run_validation_set:
+                val_ds = train.get_dataset_shard("val")
+                batch_iter_val = val_ds.iter_torch_batches(batch_size=args.batch_size)
+
+        print(f"Epoch {epoch+1} of {args.num_epochs}")
+        num_rows = 0
+        start_t = time.time()
+        num_batches = 0.0
+        total_loss = 0.0
+        for batch_idx, batch in enumerate(batch_iter):
             if not (args.use_torch or args.use_mosaic):
-                batch = batch["image"]
-            # `batch` should have tensor in `torch.Tensor` format.
-            num_rows += batch.size(dim=0)
-            if worker_rank == 0 and num_rows >= print_at:
+                num_rows += batch["image"].size(dim=0)
```

Comment:
Nit: might be nice to separate out some of these deep code blocks into helper methods but it's also fine if not.